### ISSUE TITLE
load test results for 1.3.0 on stage

### ISSUE THREAD

**Issue by pdehaan** [mozilla-services/readinglist#215]
I hit the [readinglist@1.3.0 stage server](https://readinglist.stage.mozaws.net/v1/) with 300 requests and the "Approximate Average RPS" seems crazy low: **7** on the first run and **4** on the second.

---

```ini
[loads]
hits = 100
users = 3
no-dns-resolve = true
include_file = loadtest/*.py
```

## Test 1

```
$ make test

date
Wed Mar 25 18:31:54 PDT 2015

time .venv/bin/loads-runner --config=./config/test.ini --server-url=https://readinglist.stage.mozaws.net:443 loadtest.TestBasic.test_all

Duration: 108.83 seconds
Hits: 823
Started: 2015-03-26 01:31:55.123004
Approximate Average RPS: 7
Average request time: 0.25s
Opened web sockets: 0
Bytes received via web sockets : 0

Success: 300
Errors: 0
Failures: 0


Slowest URL: https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzNTQ1NzM3fQ%3D%3D   Average Request Time: 9.070096

Stats by URLs (10 slowests):
- /v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzNTQ1NzM3fQ%3D%3D  Average request time: 9.070096
- /v1/articles/0fd2205d-b22b-48bb-92d2-8671da836d57                           Average request time: 7.312085
- /v1/articles/dae26c7a-9256-4819-8ec5-88c615d3998c                           Average request time: 0.748346
- /v1/articles/6645db23-4907-4981-862e-82515d388699                           Average request time: 0.730897
- /v1/articles/582dd526-fe64-4b06-8e9d-8ead06a3f62d                           Average request time: 0.607855
- /v1/batch                                                                   Average request time: 0.587210293478
- /v1/articles/abb48378-463f-42cb-9ebc-34fed02376f4                           Average request time: 0.552486
- /v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzNTc1NzcwfQ%3D%3D  Average request time: 0.525809
- /v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzNTI0ODM1fQ%3D%3D  Average request time: 0.522074
- /v1/articles/e35aea48-0506-4f3e-b5fd-afb3d2ebb462                           Average request time: 0.521437

Custom metrics:
- 200 : 1193
- 201 : 628
- 404 : 80
- 409 : 1
- archive : 4
- batch_archive : 9
- batch_count : 25
- batch_create : 21
- batch_delete : 4
- batch_read_further : 33
- create : 7
- create_conflict : 2
- delete : 2
- filter_sort : 31
- list_archived : 5
- list_continuated_pagination : 26
- list_deleted : 16
- mark_as_read : 16
- poll_changes : 40
- read_further : 35
- update : 23
- update_conflict : 1

109.54 real
  4.40 user
  0.78 sys
```

---

## Test 2:
```
$ make test
date
Wed Mar 25 18:36:37 PDT 2015

time .venv/bin/loads-runner --config=./config/test.ini --server-url=https://readinglist.stage.mozaws.net:443 loadtest.TestBasic.test_all

Duration: 207.24 seconds
Hits: 917
Started: 2015-03-26 01:36:37.651085
Approximate Average RPS: 4
Average request time: 0.28s
Opened web sockets: 0
Bytes received via web sockets : 0

Success: 300
Errors: 0
Failures: 0


Slowest URL: https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzODc2Mjg3fQ%3D%3D   Average Request Time: 6.071172

Stats by URLs (10 slowests):
- /v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzODc2Mjg3fQ%3D%3D  Average request time: 6.071172
- /v1/articles?_since=1427333816543                                           Average request time: 3.322751
- /v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzODEyNjMzfQ%3D%3D  Average request time: 2.351526
- /v1/articles/1b8697d1-e9ed-4c3a-85c4-0f1457006a8b                           Average request time: 1.746042
- /v1/articles?_since=1427333822156&deleted=true                              Average request time: 1.2838
- /v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzODA5ODc5fQ%3D%3D  Average request time: 1.006385
- /v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzODczMTY2fQ%3D%3D  Average request time: 0.963719
- /v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzODc2MTAxfQ%3D%3D  Average request time: 0.952659
- /v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzODg0MDA2fQ%3D%3D  Average request time: 0.877776
- /v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzMzODEwMjc0fQ%3D%3D  Average request time: 0.715044

Custom metrics:
- 200 : 1156
- 201 : 608
- 404 : 120
- 409 : 5
- archive : 5
- batch_archive : 16
- batch_count : 18
- batch_create : 20
- batch_delete : 6
- batch_read_further : 32
- create : 7
- create_conflict : 8
- delete : 5
- filter_sort : 29
- list_archived : 8
- list_continuated_pagination : 27
- list_deleted : 19
- mark_as_read : 15
- poll_changes : 30
- read_further : 36
- update : 14
- update_conflict : 5

207.85 real
  5.73 user
  0.89 sys
```

**Comment by leplatrem** [comment]
With the *Average Request Time* of the slowest URLs (9 sec!), no wonder the RPS falls down...

**Comment by pdehaan** [comment]
@leplatrem It's hard to tell if the 9s and 7s requests were outliers, or a perf regression.

I can start a bigger load test against stage right now and it should be done in a couple hours.
Plus, tomorrow morning I'll look through the Travis builds and check the load test results from each of the builds over the past 7 days and see if they avg slowest requests are in the <1s range or not.

**Comment by pdehaan** [comment]
Ran another 2 quick 300 request tests and avg request time was 0.66s - 0.81s!!

```
Duration: 70.95 seconds
Hits: 873
Started: 2015-03-26 07:58:39.213344
Approximate Average RPS: 12
Average request time: 0.17s
Opened web sockets: 0
Bytes received via web sockets : 0

Success: 300
Errors: 0
Failures: 0


Slowest URL: https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427356721611  Average Request Time: 0.8097635

Stats by URLs (10 slowests):
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427356721611                                           Average request time: 0.809763
- https://readinglist.stage.mozaws.net:443/v1/articles/c3b56463-0d13-49c5-847b-38fe47f75b9f                           Average request time: 0.720677
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzU2NzQ5NzQzfQ%3D%3D  Average request time: 0.519468
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427356749413                                           Average request time: 0.516903
- https://readinglist.stage.mozaws.net:443/v1/articles/62f029d3-2cc9-4b6f-8861-d02cd95abed8                           Average request time: 0.511138
- https://readinglist.stage.mozaws.net:443/v1/batch                                                                   Average request time: 0.466576817073
- https://readinglist.stage.mozaws.net:443/v1/articles/fed8e3da-ca52-43c7-aad7-8d34126d6bce                           Average request time: 0.465595
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427356721401&deleted=true                              Average request time: 0.459725
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzU2NzM5MzYyfQ%3D%3D  Average request time: 0.297232
- https://readinglist.stage.mozaws.net:443/v1/articles/a68b459f-c599-4f23-a086-94b0a44f8a5d                           Average request time: 0.279831
```

```
Duration: 89.02 seconds
Hits: 1135
Started: 2015-03-26 08:00:06.633263
Approximate Average RPS: 12
Average request time: 0.15s
Opened web sockets: 0
Bytes received via web sockets : 0

Success: 300
Errors: 0
Failures: 0


Slowest URL: https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzU2ODQ3MTQ0fQ%3D%3D   Average Request Time: 0.660923

Stats by URLs (10 slowests):
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzU2ODQ3MTQ0fQ%3D%3D  Average request time: 0.660923
- https://readinglist.stage.mozaws.net:443/v1/articles/5121fade-49ef-47f6-ae04-653a945f9d7d                           Average request time: 0.548859
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427356814290                                           Average request time: 0.493241
- https://readinglist.stage.mozaws.net:443/v1/articles/ff19f6ef-b13f-4158-b579-8c3d50cc27d8                           Average request time: 0.481153
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzU2ODA5NTM5fQ%3D%3D  Average request time: 0.459683
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzU2ODEyODc1fQ%3D%3D  Average request time: 0.456282
- https://readinglist.stage.mozaws.net:443/v1/batch                                                                   Average request time: 0.442203630952
- https://readinglist.stage.mozaws.net:443/v1/articles/e9e95380-5a8c-4c1a-85a4-5f45401ddbab                           Average request time: 0.430853
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427356814765&deleted=true                              Average request time: 0.395106
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427356808620                                           Average request time: 0.389979
```

**Comment by pdehaan** [comment]
Results from a 3600sec (1hr) load test:

**TL;DR:**
Approximate Average RPS: **26**
Average request time: **0.58s**
Success: **15844** (0 errors/failures)
Top 10 slowest URLs ranged from **4.9s** to **7.1s**. (!!)

---

```sh
$ http GET https://readinglist.stage.mozaws.net/v1/

HTTP/1.1 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: DNT,X-Mx-ReqToken,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Authorization,X-Conditions-Accepted
Access-Control-Allow-Methods: GET, POST, OPTIONS
Access-Control-Expose-Headers: Backoff, Retry-After, Alert
Access-Control-Max-Age: 1728000
Backoff: None
Connection: keep-alive
Content-Length: 146
Content-Type: application/json; charset=UTF-8
Date: Thu, 26 Mar 2015 08:07:42 GMT

{
    "documentation": "https://readinglist.readthedocs.org/",
    "hello": "readinglist",
    "url": "https://readinglist.stage.mozaws.net",
    "version": "1.3.0"
}
```

```sh
$ make megabench
% date
Thu Mar 26 01:07:53 PDT 2015

% cat ./config/megabench.ini
[loads]
users = 20
duration = 3600
include_file = loadtest
# agents = 5
# detach = true
# observer = irc
# ssh = ubuntu@loads.services.mozilla.com
no-dns-resolve = true
# broker = tcp://172.31.44.86:7780

% time .venv/bin/loads-runner --config=./config/megabench.ini --user-id=pdehaan --server-url=https://readinglist.stage.mozaws.net:443 loadtest.TestBasic.test_all

Duration: 3600.04 seconds
Hits: 95278
Started: 2015-03-26 08:07:53.445741
Approximate Average RPS: 26
Average request time: 0.58s
Opened web sockets: 0
Bytes received via web sockets : 0

Success: 15844
Errors: 0
Failures: 0


Slowest URL: https://readinglist.stage.mozaws.net:443/v1/articles/8c63fe88-1fe2-44c0-a7e1-49900c1cdc03 	Average Request Time: 7.090854

Stats by URLs (10 slowests):
- https://readinglist.stage.mozaws.net:443/v1/articles/8c63fe88-1fe2-44c0-a7e1-49900c1cdc03                         	Average request time: 7.090854	Hits success rate: 1.0
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427359024410&deleted=true                            	Average request time: 5.742572	Hits success rate: 1.0
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427357424984&deleted=true                            	Average request time: 5.658728	Hits success rate: 1.0
- https://readinglist.stage.mozaws.net:443/v1/articles/125dd7d2-d35a-4f4e-8dcc-7ab1b2aaa482                         	Average request time: 5.252254	Hits success rate: 1.0
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427359034745                                         	Average request time: 5.180965	Hits success rate: 1.0
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzU5OTAxMzEyfQ%3D%3D	Average request time: 5.121212	Hits success rate: 1.0
- https://readinglist.stage.mozaws.net:443/v1/articles/4f595026-88d8-4127-a004-05e3d2d0bea3                         	Average request time: 5.104203	Hits success rate: 1.0
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzU5NTYyNjU5fQ%3D%3D	Average request time: 5.055346	Hits success rate: 1.0
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzU4MTIzMjkxfQ%3D%3D	Average request time: 5.021331	Hits success rate: 1.0
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzU5MTA2MDI3fQ%3D%3D	Average request time: 4.870596	Hits success rate: 1.0

Custom metrics:
- 200 : 61124
- 201 : 28477
- 404 : 3820
- 409 : 213
- archive : 205
- batch_archive : 628
- batch_count : 1048
- batch_create : 1084
- batch_delete : 191
- batch_read_further : 1738
- create : 403
- create_conflict : 191
- delete : 200
- filter_sort : 1289
- list_archived : 387
- list_continuated_pagination : 1733
- list_deleted : 924
- mark_as_read : 841
- poll_changes : 1939
- read_further : 1751
- update : 1092
- update_conflict : 213

     7340.67 real      4467.34 user       153.53 sys

say "build complete"
```

**Comment by pdehaan** [comment]
More load test results...

```

Duration: 3600.05 seconds
Hits: 98146
Started: 2015-03-26 18:02:07.424930
Approximate Average RPS: 27
Average request time: 0.64s
Opened web sockets: 0
Bytes received via web sockets : 0

Success: 16201
Errors: 0
Failures: 0


Slowest URL: https://readinglist.stage.mozaws.net:443/v1/articles/efabab6b-acda-4b23-af07-352955485d03  Average Request Time: 5.14018

Stats by URLs (10 slowests):
- https://readinglist.stage.mozaws.net:443/v1/articles/efabab6b-acda-4b23-af07-352955485d03                           Average request time: 5.14018
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427393087896&deleted=true                              Average request time: 5.12489
- https://readinglist.stage.mozaws.net:443/v1/articles/5a365805-11fe-44db-9c77-d21f768a45a1                           Average request time: 4.93718
- https://readinglist.stage.mozaws.net:443/v1/articles/d85c18a7-23a3-45bc-8742-29dd16f1db1f                           Average request time: 4.84539
- https://readinglist.stage.mozaws.net:443/v1/articles/d16b1459-87e4-449e-a54b-e6b74c3e68ab                           Average request time: 4.60684
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3Mzk0NTIzNDk2fQ%3D%3D  Average request time: 4.55437
- https://readinglist.stage.mozaws.net:443/v1/articles/9f5ef56e-baa9-40ae-9538-57086f0493dd                           Average request time: 4.50062
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3Mzk0ODc1NzAwfQ%3D%3D  Average request time: 4.47970
- https://readinglist.stage.mozaws.net:443/v1/articles?_since=1427393753368                                           Average request time: 4.33685
- https://readinglist.stage.mozaws.net/v1/articles?_limit=20&_token=eyJsYXN0X21vZGlmaWVkIjoxNDI3MzkzNDM2OTgwfQ%3D%3D  Average request time: 4.27020

Custom metrics:
- 200 : 61195
- 201 : 28303
- 404 : 4620
- 409 : 202
- archive : 205
- batch_archive : 651
- batch_count : 1085
- batch_create : 1077
- batch_delete : 231
- batch_read_further : 1713
- create : 450
- create_conflict : 227
- delete : 221
- filter_sort : 1331
- list_archived : 432
- list_continuated_pagination : 1809
- list_deleted : 859
- mark_as_read : 854
- poll_changes : 1949
- read_further : 1755
- update : 1166
- update_conflict : 202

7331.24 real
4395.88 user
 125.49 sys
```

**Comment by ametaireau** [comment]
My understanding is that this load test was run after we tweaked stage to use gevent (~17h our time), and  @pdehaan pingued me with these last results at approx 21h30 our time, so I guess it was run with gevent enabled on stage.

**Comment by pdehaan** [comment]
I think this is old enough and we've tweaked enough settings that we can close this and open new bugs if we still see this in stage with the new stacks.
